### PR TITLE
Fix potential security issue(s), documentation on minetest.deserialize()

### PR DESCRIPTION
**Pull request by luk3yx**
- Goal of the PR

This PR makes the documentation on minetest.deserialize() less misleading and prevents functions from (easily) accessing the global environment.

- How does the PR work?

When loadstring is called inside `minetest.deserialize()` code, the resulting function will have an empty environment.

- Does it resolve any reported issue?

No.

## To do

This PR is a ready for review.

## How to test

The following code will no longer work:
```lua
minetest.deserialize('loadstring("minetest.after(0, error)")()')
```

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -5214,10 +5214,14 @@ Misc.
...
       into string form readable by `minetest.deserialize`
     * Example: `serialize({foo='bar'})`, returns `'return { ["foo"] = "bar" }'`
-* `minetest.deserialize(string)`: returns a table
+* `minetest.deserialize(string[, safe])`: returns a table
     * Convert a string returned by `minetest.deserialize` into a table
     * `string` is loaded in an empty sandbox environment.
-    * Will load functions, but they cannot access the global environment.
+    * Will load functions if safe is false or omitted. Although these functions
+      cannot directly access the global environment, they could bypass this
+      restriction with maliciously crafted Lua bytecode.
```
Comment (sfan5) [inline doc/lua_api.txt]: Is that so? Mod security prevents loading of bytecode.

**Comment by sfan5** [comment]
Hm. Since mod security prevents loading of bytecode, function (de-)serialization will not have worked before either.
Might be better to deprecate or remove that entirely.

**Comment by sorcerykid** [comment]
Why does the description for `minetest.deserialize()` state "Convert a string returned by `minetest.deserialize` into a table". Why would this method accept a string returned by itself? What would be the benefit of recursive deserialization?

**Comment by luk3yx** [comment]
> Hm. Since mod security prevents loading of bytecode, function (de-)serialization will not have worked before either.
> Might be better to deprecate or remove that entirely.

It does work (and `minetest.serialize` exploits this):

```lua
minetest.deserialize([[loadstring("\27LJ\2\0\6=stdin2\0\0\2\0\2\0\4\5\1\0006\0\0\0'\1\1\0B\0\2\1K\0\1\0\17Hello world!\nprint\0\0\0\0\0\0")]]
```

The bytecode limitation is nothing more than [security theatre](https://en.wikipedia.org/wiki/Security_theater) without the `safe` parameter.

> Why does the description for `minetest.deserialize()` state "Convert a string returned by `minetest.deserialize` into a table". Why would this method accept a string returned by itself? What would be the benefit of recursive deserialization?

Fixed.

**Comment by Wuzzy2** [comment]
Oooff, this function can actually execute code??! This sounds like a critical security vulnerability waiting to happen.

File: builtin/common/serialize.lua [hunk 2]
```diff
@@ -195,7 +198,7 @@ function core.deserialize(str, safe)
 	end
 	local f, err = loadstring(str)
 	if not f then return nil, err end
-	setfenv(f, safe and safe_env or env)
+	setfenv(f, safe and {loadstring = function() end} or env)
```
Comment (sfan5) [inline builtin/common/serialize.lua]: Actually, it's better to move this back into a `local safe_env = {` above so the function is not unnecessarily recreated on each call.

Comment (luk3yx) [reply builtin/common/serialize.lua]: Added.

**Comment by luk3yx** [comment]
Should it recreate the unsafe environment (`env`) every time?

File: builtin/common/serialize.lua (hunk 2 above)
Comment (sfan5) [reply builtin/common/serialize.lua]: `env` needs to be newly constructed on every invocation.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -5214,10 +5214,15 @@ Misc.
...
+* `minetest.deserialize(string[, safe])`: returns a table
+    * Convert a string returned by `minetest.serialize` into a table
     * `string` is loaded in an empty sandbox environment.
-    * Will load functions, but they cannot access the global environment.
+    * Will load functions if safe is false or omitted. Although these functions
+      cannot directly access the global environment, they could bypass this
+      restriction with maliciously crafted Lua bytecode if mod security is
+      disabled.
+    * This function should not be used on untrusted data, regardless of the
+      value of safe.
```
Comment (rubenwardy) [inline doc/lua_api.txt]: > 
```suggestion
      value of safe. It is perfectly fine to serialize then deserialize user-provided data,
      but directly providing user-input to deserialize is not safe.
```

File: doc/lua_api.txt [hunk 4]
```diff
@@ -5214,10 +5214,16 @@ Misc.
...
+    * Convert a string returned by `minetest.serialize` into a table
     * `string` is loaded in an empty sandbox environment.
-    * Will load functions, but they cannot access the global environment.
+    * Will load functions if safe is false or omitted. Although these functions
+      cannot directly access the global environment, they could bypass this
+      restriction with maliciously crafted Lua bytecode if mod security is
+      disabled.
+    * This function should not be used on untrusted data, regardless of the
+     value of `safe`. It is fine to serialize then deserialize user-provided
+     data, but directly providing user-input to deserialize is not safe.
```
Comment (sfan5) [inline doc/lua_api.txt]: I kinda don't like the wording here

```suggestion
     data, but directly providing user input to deserialize is always unsafe.
```

**Comment by rubenwardy** [comment]
I'm finding it hard to think about edge cases to test with this. Here are two tests, though:

```lua.diff
diff --git a/builtin/common/tests/serialize_spec.lua b/builtin/common/tests/serialize_spec.lua
index 321d2766a..e1a8a5c3d 100644
--- a/builtin/common/tests/serialize_spec.lua
+++ b/builtin/common/tests/serialize_spec.lua
@@ -1,6 +1,6 @@
 _G.core = {}
 
-_G.setfenv = function() end
+_G.setfenv = require 'busted.compatibility'.setfenv
 
 dofile("builtin/common/serialize.lua")
 
@@ -25,4 +25,30 @@ describe("serialize", function()
                local test_out = core.deserialize(core.serialize(test_in))
                assert.same(test_in, test_out)
        end)
+
+       it("strips functions in safe mode", function()
+               local test_in = {
+                       func = function(a, b)
+                               error("test")
+                       end,
+                       foo = "bar"
+               }
+
+               local str = core.serialize(test_in)
+               assert.not_nil(str:find("loadstring"))
+
+               local test_out = core.deserialize(str, true)
+               assert.is_nil(test_out.func)
+               assert.equals(test_out.foo, "bar")
+       end)
+
+       it("safe mode can't be blocked", function()
+               local str = [[
+                       return {["func"] = loadstring("while true do end"), ["foo"] = "bar" }
+               ]]
+
+               local test_out = core.deserialize(str, true)
+               assert.is_nil(test_out.func)
+               assert.equals(test_out.foo, "bar")
+       end)
 end)
```

**Comment by sfan5** [comment]
The first test case is fine, though it may be better to just provide the string instead of calling serialize.

The second does not make sense. It is always possible to "block" the deserialization by providing `"while true do end"` as input, safe mode doesn't fix this.

**Comment by rubenwardy** [comment]
> The first test case is fine, though it may be better to just provide the string instead of calling serialize.

The string ends up like this, including random characters:

```lua
return {["func"] = loadstring("\27LuaS\0\25�\13\
\26\
\4\8\4\8\8xV\0\0\0\0\0\0\0\0\0\0\0(w@\1)@builtin/common/tests/serialize_spec.lua\31\0\0\0!\0\0\0\2\0\4\4\0\0\0�\0@\0�@\0\0�@\0\1&\0�\0\2\0\0\0\4\6error\4\5test\1\0\0\0\0\0\0\0\0\0\4\0\0\0 \0\0\0 \0\0\0 \0\0\0!\0\0\0\2\0\0\0\2a\0\0\0\0\4\0\0\0\2b\0\0\0\0\4\0\0\0\1\0\0\0\5_ENV"), ["foo"] = "bar"}
```

Using serialize feels cleaner

**Comment by sfan5** [comment]
`return {["func"] = loadstring(""), ["foo"] = "bar"}` would work just as well for a test string
